kernel/src/drivers/terminal/mod.rs

```diff
@@ -0,0 +1,316 @@
+use core::ops::Add;
+use core::fmt;
+use core::result::Result;
+use spin::Mutex;
+use drivers::vga::{self, VgaWriter};
+use color::Color;
+
+pub static WRITER: Mutex<VgaWriter> = Mutex::new(VgaWriter::new());
+pub static STDOUT: Mutex<TextArea<VgaWriter>> = Mutex::new(TextArea::new(&WRITER, Point::new(0, 0), Point::new(vga::RESOLUTION_X, vga::RESOLUTION_Y)));
+
+/// A general error terminal error
+#[derive(Debug)]
+pub enum TerminalWriteError<T: TerminalWriter> {
+    OutputError(T::Error),
+    BackspaceUnsupported,
+    BackspaceUnavailable,
+    OutOfBounds(Point),
+}
+
+/// A terminal with a constant stream of characters.
+#[allow(dead_code)] // to be used as API
+pub struct StreamTerminal<'a, T: TerminalWriter + 'a> {
+    writer: &'a Mutex<T>,
+    color: TerminalColor,
+}
+
+#[allow(dead_code)] // to be used as API
+impl<'a, T: TerminalWriter + 'a> StreamTerminal<'a, T> {
+    pub const fn new(writer: &'a Mutex<T>) -> Self {
+        StreamTerminal {
+            writer: writer,
+            color: TerminalColor::new(Color::White, Color::Black),
+        }
+    }
+}
+
+impl<'a, T: TerminalWriter + 'a> Terminal<'a, T> for StreamTerminal<'a, T> {
+    fn set_color(&mut self, color: TerminalColor) -> Result<(), TerminalWriteError<T>> {
+        self.color = color;
```

Comment:
This should check whether the color is supported, no?

Reply:
Support for colors would be handled in the `TerminalWriter`.

Reply:
Then what is the point of having unsupported color error if we cannot catch it from here (T::Error is generic so defaults to write error)? That could work if every T::Error has impl From for TerminalOutputError which selects appropriately... Also, that would have to check *every single time*, which is kinda inefficient...